pkg/gardenadm/cmd/discover/discover.go

```diff
@@ -99,3 +164,83 @@ func readShoot(fs afero.Afero, manifestPath string) (*gardencorev1beta1.Shoot, e
+		secretBinding := &gardencorev1beta1.SecretBinding{ObjectMeta: metav1.ObjectMeta{Name: *shoot.Spec.SecretBindingName, Namespace: shoot.Namespace}}
+		return secretBinding, c.Get(ctx, client.ObjectKeyFromObject(secretBinding), secretBinding)
+
+	case shoot.Spec.CredentialsBindingName != nil:
+		credentialsBinding := &securityv1alpha1.CredentialsBinding{ObjectMeta: metav1.ObjectMeta{Name: *shoot.Spec.CredentialsBindingName, Namespace: shoot.Namespace}}
+		return credentialsBinding, c.Get(ctx, client.ObjectKeyFromObject(credentialsBinding), credentialsBinding)
+
+	default:
+		return nil, nil
+	}
+}
+
+func requiredExtensions(ctx context.Context, c client.Client, shoot *gardencorev1beta1.Shoot, runsControlPlane bool) ([]botanist.Extension, error) {
+	resources := gardenadm.Resources{Shoot: shoot}
+
+	controllerRegistrationList := &gardencorev1beta1.ControllerRegistrationList{}
+	if err := c.List(ctx, controllerRegistrationList); err != nil {
+		return nil, fmt.Errorf("failed listing controllerRegistrations: %w", err)
+	}
+	controllerDeploymentList := &gardencorev1.ControllerDeploymentList{}
+	if err := c.List(ctx, controllerDeploymentList); err != nil {
+		return nil, fmt.Errorf("failed listing controllerDeployments: %w", err)
+	}
+
+	if err := meta.EachListItem(controllerRegistrationList, func(obj runtime.Object) error {
+		resources.ControllerRegistrations = append(resources.ControllerRegistrations, obj.(*gardencorev1beta1.ControllerRegistration))
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed adding ControllerRegistrations: %w", err)
+	}
+
+	if err := meta.EachListItem(controllerDeploymentList, func(obj runtime.Object) error {
+		resources.ControllerDeployments = append(resources.ControllerDeployments, obj.(*gardencorev1.ControllerDeployment))
+		return nil
+	}); err != nil {
+		return nil, fmt.Errorf("failed adding ControllerDeployments: %w", err)
+	}
+
+	return botanist.ComputeExtensions(resources, runsControlPlane)
+}
+
+func getAndExportObject(ctx context.Context, c client.Client, fs afero.Afero, opts *Options, kind string, obj client.Object) error {
+	if err := c.Get(ctx, client.ObjectKeyFromObject(obj), obj); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed getting %s %q: %w", kind, client.ObjectKeyFromObject(obj), err)
+		}
+		opts.Log.V(1).Info("Object not found in garden cluster", "kind", kind, "obj", client.ObjectKeyFromObject(obj))
+		return nil
+	}
+
+	resetObject(obj)
+
+	objYAML, err := kubernetesutils.Serialize(obj, kubernetes.GardenScheme)
+	if err != nil {
+		return fmt.Errorf("failed serializing %T %q: %w", obj, client.ObjectKeyFromObject(obj), err)
+	}
+
+	path := filepath.Join(opts.ConfigDir, fmt.Sprintf("%s-%s.yaml", strings.ToLower(kind), obj.GetName()))
+	if err := fs.WriteFile(path, []byte(objYAML), 0644); err != nil {
+		return fmt.Errorf("failed writing file to %s: %w", path, err)
+	}
+
+	fmt.Fprintf(opts.Out, "Exported %s to %s\n", kind, path)
```

Comment:
Should we include the name here explicitly as well or do you think it is enough that it shows up in the path?

For example:

```suggestion
	fmt.Fprintf(opts.Out, "Exported %s/%s to %s\n", kind, obj.GetName(), path)
```